development-docs/DEBUGGING.md

```diff
@@ -0,0 +1,188 @@
+# Debugging Guide for Strimzi
+
+This guide primarily explains how to perform remote debugging of code deployed to Kubernetes with strimzi/operator image, and the strimzi/kafka images.
+
+<!-- TOC depthFrom:2 -->
+
+- [Activating remote debugging using the agent](#activating-remote-debugging-using-the-agent)
+- [Remote debugging the Strimzi Cluster Operator](#remote-debugging-the-strimzi-cluster-operator)
+- [Remote debugging the Kafka Broker](#remote-debugging-the-kafka-broker)
+- [Using the IntelliJ IDEA for debugging](#using-the-intellij-idea-for-debugging)
+  - [Setting the breakpoints](#setting-the-breakpoints)
+  - [Evaluating expressions when suspended](#evaluating-expressions-when-suspended)
+  - [Resolving code version mismatches](#resolving-code-version-mismatches)
+
+<!-- /TOC -->
+
+## Activating remote debugging using the agent
+
+Java runtime (the JVM) comes with an agent that provides support for remote debugging. Using you Java IDE you can connect to a remote process over a TCP connection, and perform debugging actions like setting breakpoints, suspending threads, gathering stack and heap info, and evaluating arbitrary java code inside the running JVM.
+
+In order to activate remote debugging the java process needs to be run with an additional argument like the following:
+
+    -agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005
+
+This starts a TCP server on localhost using the specified port (5005 in this case). Connection is only possible from the localhost.
+Because we used `suspend=y` the start up of the JVM will stall, and wait for the debug client to connect.
+All the popular Java IDEs support remote debugging.
+
+
+## Remote debugging the Strimzi Cluster Operator
+
+Strimzi Cluster Operator is installed through Deployment definition in `install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml`.
+
+We can enable remote debugging by adding the `JAVA_OPTS` environment variable to `strimzi-cluster-operator` container in this file:
+
+          env:
+            ...
+            - name: JAVA_OPTS
+              value: "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005"
+
+
+Let's also increase the livenessProbe timeout:
+
+          livenessProbe:
+            initialDelaySeconds: 3600
+            timeoutSeconds: 3600
+
+Apply the modified file:
+
+    kubectl apply -f install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml
+
+We also need to open the port for localhost access:
+
+    kubectl port-forward $(oc get pod | grep strimzi-cluster-operator | awk '{printf $1}') 5006:5005
```

Comment:
don't think that changing the local port is necessary, it could just confuse user that now should attach the debugger on localhost:5006 instead of 5005

Reply:
Good catch, not supposed to be like this - just I locally used it like this because I had two port-forwards of 5005 open at the same time.